kani-compiler/src/codegen_cprover_gotoc/reachability.rs

```diff
@@ -454,67 +453,111 @@ fn should_codegen_locally<'tcx>(tcx: TyCtxt<'tcx>, instance: &Instance<'tcx>) ->
+        // the conversion.
+        match (&src_inner_ty.kind(), &dst_inner_ty.kind()) {
+            (&ty::Adt(src_def, src_substs), &ty::Adt(dst_def, dst_substs))
+                if !src_def.is_box() || !dst_def.is_box() =>
+            {
+                // Handle smart pointers by using CustomCoerceUnsized to find the field being
+                // coerced.
+                assert_eq!(src_def, dst_def);
+                let src_fields = &src_def.non_enum_variant().fields;
+                let dst_fields = &dst_def.non_enum_variant().fields;
+                assert_eq!(src_fields.len(), dst_fields.len());
+
+                let CustomCoerceUnsized::Struct(coerce_index) =
+                    custom_coerce_unsize_info(tcx, src_inner_ty, dst_inner_ty);
+                assert!(coerce_index < src_fields.len());
+
+                src_inner_ty = src_fields[coerce_index].ty(tcx, src_substs);
+                dst_inner_ty = dst_fields[coerce_index].ty(tcx, dst_substs);
+            }
+            // Base case is always a pointer (Box, raw_pointer or reference).
+            _ => break (extract_pointer(src_inner_ty), extract_pointer(dst_inner_ty)),
+        }
+    };
 
-    trace!(?dst_ty_inner, ?src_ty_inner, "find_trait_conversion");
-    (has_vtable_metadata(dst_ty_inner) && !has_vtable_metadata(src_ty_inner)).then(|| {
-        let param_env = ParamEnv::reveal_all();
-        tcx.struct_lockstep_tails_erasing_lifetimes(src_ty_inner, dst_ty_inner, param_env)
-    })
+    if has_vtable_metadata(tcx, src_inner_ty) {
+        (src_inner_ty, dst_inner_ty)
+    } else {
+        tcx.struct_lockstep_tails_erasing_lifetimes(
+            src_inner_ty,
+            dst_inner_ty,
+            ParamEnv::reveal_all(),
+        )
+    }
 }
 
-/// This function extracts the pointee type of a regular pointer and std smart pointers.
-///
-/// TODO: Refactor this to use `CustomCoerceUnsized` logic which includes custom smart pointers.
-/// <https://github.com/model-checking/kani/issues/1694>
-///
-/// E.g.: For `Rc<dyn T>` where the Rc definition is:
-/// ```
-/// pub struct Rc<T: ?Sized> {
-///    ptr: NonNull<RcBox<T>>,
-///    phantom: PhantomData<RcBox<T>>,
-/// }
-///
-/// pub struct NonNull<T: ?Sized> {
-///    pointer: *const T,
-/// }
-/// ```
-///
-/// This function will return `pointer: *const T`.
-fn extract_pointer<'tcx>(tcx: TyCtxt<'tcx>, typ: Ty<'tcx>) -> Ty<'tcx> {
+/// Extract pointer from builtin pointer types.
+fn extract_pointer(typ: Ty) -> Ty {
     if let Some(TypeAndMut { ty, .. }) = typ.builtin_deref(true) {
         ty
     } else {
-        ReceiverIter { tcx, curr: typ }.last().unwrap().1
+        unreachable!("Expected pointer type, found: {:?}", typ);
+    }
+}
+
+/// Check if the type has metadata.
+fn has_vtable_metadata<'tcx>(tcx: TyCtxt<'tcx>, mir_type: Ty<'tcx>) -> bool {
+    let (metadata, _) = mir_type.ptr_metadata_ty(tcx, |ty| ty);
+    metadata != tcx.types.unit && metadata != tcx.types.usize
```

Comment:
I know these are currently the only three kinds of metadata, but is it possible to future proof this in case new types are invented?

Reply:
what do you have in mind?

Reply:
Looking back, I think this check might be redundant. Let me see if I can remove this code for now. FYI, we do have the exact same check here:

https://github.com/model-checking/kani/blob/19add43fd2f2f3cbc2f6ef25fef4d12d1b0dc39d/kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs#L1968-L1971